### ISSUE TITLE
[Pyright reportGeneralTypeIssues] Type "ForwardRef" cannot be assigned to type "Type[Foo]"

### ISSUE THREAD

**Issue by SirJls** [microsoft/pyright#1960]
**Describe the bug**
Forward declaration not working properly with pyright.

**To Reproduce**
Copy the following snippet in your favorite editor, with the assumption that pyright is properly setup:
```
from typing import ForwardRef, Optional

Foo = ForwardRef('Foo')


class Foo:
    def __init__(self, i: int, foo: Optional[Foo] = None):
        self.a: int = i
        self.b: Optional[Foo] = foo



print(Foo(4, Foo(4)).a)
print(Foo(1, Foo(10)).a)
```

**Expected behavior**
The following errors are not raised by pyright:
```
snippet.py|3 col 1 error| [Pyright reportGeneralTypeIssues] Declaration "Foo" is obscured by a declaration of the same name [E]
snippet.py|3 col 7 error| [Pyright reportGeneralTypeIssues] Expression of type "ForwardRef" cannot be assigned to declared type "T
```

**Screenshots or Code**
https://i.imgur.com/idcInlF.png 
[](url)

**VS Code extension or command-line**
Command-line with version: 1.1.147 (latest at the moment of writing).

**Additional context**
-

**Comment by erictraut** [comment]
This isn't how forward references work in the Python type system. `ForwardRef` is a runtime class used to introspect forward references, but it's not intended to be used in type aliases or type annotations.

The proper way to use forward references is to enclose the annotation in quotes:

```python
from typing import Optional

class Foo:
    def __init__(self, i: int, foo: "Optional[Foo]" = None):
        self.a: int = i
        self.b: Optional[Foo] = foo

print(Foo(4, Foo(4)).a)
print(Foo(1, Foo(10)).a)
```

**Comment by erictraut** [comment]
I'm going to close this issue, but let us know if you have any further questions.

**Comment by SirJls** [comment]
Hi Eric,

Sorry for the late response!

I seem to have wrongly understood the usage of the ForwardRef concept in Python.
Thanks for the clarification! I even learned something in the process (which is 
always good).

So yes I am okay with the closure of this issue.

Thanks for the support!


On 09-06-21 08:27, Eric Traut wrote:
>I'm going to close this issue, but let us know if you have any further questions.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/microsoft/pyright/issues/1960#issuecomment-857801973

-- 
Sjors Sparreboom
Web: http://sjorssparreboom.com
Tel: (+31) 641 77 44 22
GPG: 9BA3 F965 3685 D349 B784 DAD8 D01E F8B5 45E6 EFAE 

"""
	Simplicity is a great virtue, but it requires hard work to achieve it and
	education to appreciate it.
	And to make matters worse: complexity sells better.
"""
